### PR TITLE
Introduce with_range on Parser

### PR DESCRIPTION
**Pull request by DRiKE**
This PR introduces a new constructor for `Parser`, similar to the existing `fn from_ref` but taking a range to initialize the parser on a subset of the octets.

It currently errors when the end of the passed range lies beyond the end of the octets, meaning passing in an open-ended range `(n..)` is not possible. Alternatively, we could accept any (open) end and limit it to the length of the octets, and not returning any error.

### CODE REVIEW

**Comment by DRiKE** [comment]
Something like this seems to work. But if we really want to support this it might be cleaner to add `fn from_ref_range_from()` and `fn from_ref_range_to()` taking a RangeFrom / RangeTo, respectively.

```rust
pub fn from_ref_with_range<R>(
    octets: &'a Octs,
    range: R
) -> Result<Self, ShortInput>
where
    Octs: AsRef<[u8]>,
    R: core::ops::RangeBounds<usize>
{
    use std::ops::Bound;
    let octets_len = octets.as_ref().len();

    let pos = match range.start_bound() {
        Bound::Unbounded => 0,
        Bound::Included(n) => {
            if *n - 1 > octets_len {
                octets_len
            } else {
                *n
            }
        }
        Bound::Excluded(_) => unreachable!() // not a thing for the
                                             // start_bound, right?
    };

    let len = match range.end_bound() {
        Bound::Unbounded => octets_len,
        Bound::Excluded(n) => {
            if *n > octets_len {
                octets_len
            } else {
                *n
            }
        }
        Bound::Included(n) => {
            if (n - 1) > octets_len {
                octets_len
            } else {
                n - 1
            }
        }
    };

    let len = if len < pos {
        pos
    } else {
        len
    };

    Ok(
        Parser {
            pos,
            len,
            octets
        }
    )
}
```

**Comment by partim** [comment]
I’d probably just call the function `with_range`.

**Comment by partim** [comment]
I wonder if it should just panic if the range is outside the octets’ full range. The use case for the function typically is where you store the range to make a parser on the fly later, so you usually have checked already. And most things that take a range on slices do panic as well, so this may even be consistent.

**Comment by DRiKE** [comment]
That sounds good, I'll adapt it and rename the fn as well.

**Comment by DRiKE** [comment]
I guess this happened partly because that's what was there in my vim buffer after a lot of undo-ing, but I also thought it sort of made sense to force to be exact (i.e. an ended `Range` and not a possibly open-ended `RangeBounds`). When you are working with a `Parser` on a stream-like bunch of octets, you might have no idea whether there is stuff before the message you are parsing, or stuff beyond the end of that message. So allowing open-ended ranges sound like a bit of a foot gun?

**Comment by partim** [comment]
I think there are legitimate use cases for using open ranges. For instance, if you want to split a sequence in two, only remembering the mid point and then use open ranges when creating parsers rather than doing some extra maths to get to the full ranges is probably safer.

So, I think allowing all kinds ranges is probably better.

**Comment by DRiKE** [comment]
There we go. It does allow ranges such as `2..=1` (equivalent to `2..2` producing a parser with 0 remaining octets), which feel strange to me because such range looks like it is decreasing. But this actually in line with e.g. slice indexing in the standard library.

**Review by partim — CHANGES_REQUESTED** [review]

**Comment by partim** [comment]
I think this would benefit from a bunch of tests, especially for the corner cases.

**Comment by DRiKE** [comment]
Tests, definitely.
With regards to `len`, that always points to what we call the 'end' of the Parser if I'm not mistaking. Should that be renamed?

**Comment by partim** [comment]
Oh, you are right. Fooled myself there, apparently. Renaming it to `end` is probably a wise move, indeed. (By the looks of it, the code itself tries to avoid using it directly and uses `remaining` instead.)

File: src/parse.rs [hunk 1]
```diff
@@ -45,6 +46,61 @@ impl<'a, Octs: ?Sized> Parser<'a, Octs> {
...
+        let pos = match range.start_bound() {
+            Bound::Unbounded => 0,
+            Bound::Included(0) => 0,
+            Bound::Included(n) => {
+                if octets_len == 0 || *n > octets_len - 1 {
+                    panic!("range start is out of range for octets")
+                }
+                *n
+            }
+            Bound::Excluded(_) => unreachable!() // not a thing for the
```
[the remark this answers is not in the document]
Comment (DRiKE) [reply src/parse.rs]: Apparently, you can specify `RangeBounds` as a tuple, so the start bound can already be exclusive:
```rust
assert_eq!(
    [1,2,3][(Bound::Excluded(0), Bound::Included(2))],
    [1,2,3][1..=2]
);
```

Fixed in the latest commit.

**Comment by DRiKE** [comment]
Fixes for your comments pushed, unit tests coming up.
And then let's do the renaming of `len` in a separate PR once we're happy with this one.

**Review by partim — APPROVED** [review]
